src/plugins/data/public/ui/saved_query_flyouts/open_saved_query_flyout.tsx

```diff
@@ -268,7 +293,19 @@ export function OpenSavedQueryFlyout({
               fill
               onClick={() => {
                 if (selectedQuery) {
-                  onQueryOpen(selectedQuery);
+                  if (
+                    // Template queries are not associated with data sources. Apply data source from current query
+                    selectedQuery.attributes.isTemplate
+                  ) {
+                    const updatedQuery: Query = {
+                      ...queryStringManager?.getQuery(),
+                      query: selectedQuery.attributes.query.query,
+                      language: selectedQuery.attributes.query.language,
+                    };
+                    queryStringManager.setQuery(updatedQuery);
```

Comment:
curious, why does this use setQuery over onQueryOpen?

Reply:
This is the line that concerns me a little and want to know more about why we need this. What are we trying to do here? 

There are 2 states at play here. Query manager state and the local searchbar state. `onQueryOpen` updates the search bar state and not the query state because the act of opening a saved query today does not update the query state until the user has decided to fire the query (similar to how you dont want typing the query to update query state until you are ready to fire the query)

Updating the query manager state here for templates changes this behaviour for users when it comes to templates, because now it means that template queries DO change the query state even if the user has not been given the opportunity to confirm that thats the query that they want to run. 

This change will not affect datasources that dont run the query by default, but for indices and index patterns, this means that opening a saved query template runs the query automatically while non templates do not.

Reply:
@ashwin-pc That's good to know. Given that context, it sounds like the way forward would involve retrieving the dataset from the query state when the saved query doesn't have a dataset (i.e., when the saved query is a template, or when the saved query is not associated with a dataset), and applying it to the saved query before calling `onQueryOpen`. Does that sound appropriate?
E.g.,
```
onClick={() => {
                if (selectedQuery) {
                  if (
                    // Template queries are not associated with data sources. Apply data source from current query
                    selectedQuery.attributes.isTemplate ||
                    // Associating a saved query with a data source is optional. If no data source is present, apply the current source.
                    !selectedQuery.attributes.query.dataset?.dataSource
                  ) {
                    selectedQuery.attributes.query.dataset = queryStringManager?.getQuery().dataset;
                  }
                  onQueryOpen(selectedQuery);
                  onClose();
                }
              }}

Reply:
Since this PR has been merged, I'll adjust this in a follow-up PR.